### ISSUE TITLE
Type checking of list operator "+" in the presence of subclassing

### ISSUE THREAD

**Issue by nnemkin** [microsoft/pyright#3576]
**Describe the bug**
The following code results in a type error:

```python
class Base:
    pass
class Derived(Base):
    pass
[Derived()] + [Base()]  # ERROR: Operator "+" not supported for types "list[Derived]" and "list[Base]"
```

**Expected behavior**
There should be no error and the type of `[Derived()] + [Base()]` should be deduced to be `list[Base]`.

I think what's missing is a type-level `list.__radd__` definition. When `list[Derived].__add__(list[Base])` fails to type check, `list[Base].__radd_(list[Derived])` should be tried and should succeed.

**VS Code extension or command-line**
VSCode Pylance v2022.6.10

**Comment by erictraut** [comment]
The current behavior is correct from a type checking perspective. This should emit an error.

The left-hand operand has an inferred type of `list[Derived]`. Because the type parameter for `list` is invariant, you cannot add a `list[Base]` to ` list[Derived]`. 

Mypy also emits the same error in this case.

If the list `[Derived()]` starts life as a `list[Base]`, then `list[Base]` can be added to it. So this works without an error:

```python
x: list[Base] = [Derived()]
x = x + [Base()]
```

**Comment by erictraut** [comment]
For more information about type annotations and type inference, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).

**Comment by erictraut** [comment]
That's not how the type system in Python works. Pyright is a standards-based type checker. If you want to change the standards, you are welcome to propose and shepherd changes. The change you're proposing is unlikely to be adopted though because it would have many other negative side effects in the type system.

If you want to use static type analysis to find bugs in your code, you will need to make some (typically small) accommodations. This is such an example.
